### PR TITLE
Uplink implants now inherit the uplink flags of the uplink they were bought from

### PR DESCRIPTION
**Pull request by RandomGamer123**
## About The Pull Request
This effectively is a new version of #59714 which instead of banning nuke ops from buying uplink implants, allows them to do so but their uplink implants will be of their type.
Uplinks implants bought from uplinks inherit the uplink flags of the uplink they were bought from.

In addition, this new version allows for nukeops to actually still get uplink implants.

## Changelog
:cl:
fix: Nuclear operatives' uplink implants now properly only show items they are allowed to buy, preventing them bypassing uplink restrictions.
/:cl:

### CODE REVIEW

**Review by coiax — CHANGES_REQUESTED** [review]
I would defer to @ninjanomnom's interpretation on how to keep the layers of component abstraction separate.

But, I think it would be better if the `UPLINK_X` flag was passed in the initialization of the uplink component. To accomplish that, you'd likely need to pass the flag to the implant?

There might be a cleaner way though.

**Comment by RandomGamer123** [comment]
The implant and all that stuff is initialised when the box's items are generated in PopulateContents, however that does not accept an argument, I maybe could get it to work by overloading the box's Initialize and PopulateContents and all the Initialize downwards to accept an argument but that also doesn't seem like a clean way of doing it.

**Comment by coiax** [comment]
Is the box used anywhere else other than this uplink entry?

**Comment by RandomGamer123** [comment]
> Is the box used anywhere else other than this uplink entry?

It's also one of the possible items in the contract kit, but other than that, nope

**Comment by coiax** [comment]
Well, instead of overriding Initialize and PopulateContents, you can just have an empty box, and spawn in the implant/implanter in the spawn_item proc.

File: code/game/objects/items/implants/implanter.dm [hunk 1]
```diff
@@ -58,6 +58,6 @@
 
 /obj/item/implanter/Initialize(mapload)
 	. = ..()
-	if(imp_type)
+	if(imp_type && (imp == null))
```
Comment (tralezab) [inline code/game/objects/items/implants/implanter.dm:61]:
```suggestion
	if(!imp && imp_type)
```

File: code/datums/components/uplink.dm [hunk 2]
```diff
@@ -254,6 +248,12 @@
 
 	var/mob/user = arguments[2]
 	owner = "[user.key]"
+	if(owner)
+		LAZYINITLIST(GLOB.uplink_purchase_logs_by_key)
```
Comment (tralezab) [inline code/datums/components/uplink.dm:252]: i wish all this was on the mind, don't you? that's not really something you need to change, just me complaining into the void. i'd even say it'd be great to have on antag datum but sadly people who aren't antag need to have uplinks as well 🤷

File: code/game/objects/items/implants/implantuplink.dm [hunk 3]
```diff
@@ -6,10 +6,11 @@
 	lefthand_file = 'icons/mob/inhands/misc/devices_lefthand.dmi'
 	righthand_file = 'icons/mob/inhands/misc/devices_righthand.dmi'
 	var/starting_tc = 0
+	var/uplink_flag = UPLINK_TRAITORS
```
Comment (tralezab) [inline code/game/objects/items/implants/implantuplink.dm:9]: document this pls

**Review by tralezab — COMMENTED** [review]
Honestly it's all up to standard codewise besides one thing you forgot to document, good stuff 👍 

I did have some nagging feelings about the actual process here but they're genuinely just bike shedding so I left most of it out

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/datums/components/uplink.dm [hunk 4]
```diff
@@ -254,6 +248,12 @@
...
+		if(GLOB.uplink_purchase_logs_by_key[owner])
+			purchase_log = GLOB.uplink_purchase_logs_by_key[owner]
+		else
+			purchase_log = new(owner, src)
```
Comment (coiax) [inline code/datums/components/uplink.dm:256]: If you move this code to `implanting`, won't this mean that uplinks that are embedded in radios or PDAs won't ever create a purchase log?

File: code/game/objects/items/implants/implantuplink.dm [hunk 5]
```diff
@@ -28,6 +29,10 @@
 	name = "implanter (uplink)"
 	imp_type = /obj/item/implant/uplink
 
+/obj/item/implanter/uplink/Initialize(mapload, uplink_flag = UPLINK_TRAITORS)
+	imp = new imp_type(src, null, uplink_flag)
```
Comment (coiax) [inline code/game/objects/items/implants/implantuplink.dm:33]: Is it possible to use named arguments here?

File: code/game/objects/items/implants/implantuplink.dm [hunk 6]
```diff
@@ -6,10 +6,11 @@
 	lefthand_file = 'icons/mob/inhands/misc/devices_lefthand.dmi'
 	righthand_file = 'icons/mob/inhands/misc/devices_righthand.dmi'
 	var/starting_tc = 0
+	var/uplink_flag = UPLINK_TRAITORS
 
-/obj/item/implant/uplink/Initialize(mapload, _owner)
+/obj/item/implant/uplink/Initialize(mapload, _owner, uplink_flag)
```
Comment (coiax) [inline code/game/objects/items/implants/implantuplink.dm:11]: While you are in the area, change `_owner` to `owner`

File: code/game/objects/items/implants/implantuplink.dm [hunk 7]
```diff
@@ -6,10 +6,11 @@
...
 	. = ..()
-	AddComponent(/datum/component/uplink, _owner, TRUE, FALSE, null, starting_tc)
+	AddComponent(/datum/component/uplink, _owner, TRUE, FALSE, uplink_flag, starting_tc)
```
Comment (coiax) [inline code/game/objects/items/implants/implantuplink.dm:13]: Use named arguments here.

File: code/datums/components/uplink.dm (hunk 4 above)
Comment (RandomGamer123) [reply code/datums/components/uplink.dm:256]: Hmm, actually i think after I fixed a bug causing owner to not propagate I think just testing for if(owner) should be sufficient so I can move it back up

File: code/game/objects/items/implants/implantuplink.dm (hunk 5 above)
Comment (RandomGamer123) [reply code/game/objects/items/implants/implantuplink.dm]: It does not seem to be possible.

**Comment by RandomGamer123** [comment]
Wait what's the difference between 689bc46 and f76e111 *sigh* I'll look into this soon

**Comment by RandomGamer123** [comment]
Oh I reverted f76e111 instead of 04e6dab

**Pull request merged**